File: 28week/LeetCode-75-Sort-Colors.py
class Solution:

    # [2,0,2,1,1,0]
    def sortColors(self, nums: list[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        def is_heap(index):
            if index == 0:
                return True
            if nums[index] <= nums[(index+1) // 2 - 1]:
                return True
            return False

        def down_heap(index):
            nums[index], nums[(index+1) // 2 - 1] = nums[(index+1) // 2 - 1], nums[index]
            if not is_heap((index+1) // 2 - 1):
                down_heap((index+1) // 2 - 1)

        def heap_pop(index):
            if index == len(nums)-1:
                return
            nums[0], nums[-(index+1)] = nums[-(index+1)], nums[0]
            up_heap(len(nums) - index)

        def up_heap(final_index):
            i = 1
            while i*2 < final_index:
                if i*2+1 < final_index:
                    if nums[i - 1] < nums[i*2 - 1] or nums[i - 1] < nums[i*2]:
                        if nums[i*2 - 1] < nums[i*2]:
                            nums[i - 1], nums[i*2] = nums[i*2], nums[i - 1]
                            i = i*2 + 1
                        else:
                            nums[i - 1], nums[i*2-1] = nums[i*2-1], nums[i - 1]
                            i = i*2
                    else:
                        break
                else:
                    if nums[i - 1] < nums[i*2 - 1]:
                        nums[i - 1], nums[i*2 - 1] = nums[i*2 - 1], nums[i - 1]
                        i = i*2
                    else:
                        break
        # heap push
        # [2,0,2,1,1,0] -> [2,1,2,0,1,0]
        for i in range(len(nums)):
            if not is_heap(i):
                down_heap(i)

        # nums.reverse()
        #print(nums)

        # heap 
        # [2,1,2,0,1,0] -> [0,0,1,1,2,2]
        for i in range(len(nums)):
            heap_pop(i)

        print(nums)

File: 5월5주차/leetcode-75-sort colors.cpp
//in-place sort

//two pointer
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int i0 = 0, i1=0, i2=nums.size()-1;
        while (i1 <= i2) {
            if (nums[i1] == 0) {
                swap(nums[i0++], nums[i1]);
                i1++;
            }
            else if (nums[i1] == 2) {
                swap(nums[i1], nums[i2--]);
            }
            else i1++;
        }
    }
};

//bubble sort
class Solution {
public:
    void sortColors(vector<int>& nums) {
        for (int i=nums.size()-1; i>0; i--) {
            for (int j=0; j<i; j++) {
                if (nums[j] > nums[j+1]) swap(nums[j], nums[j+1]);
            }
        }
    }
};